### DISCUSSION TITLE
From vert.x Future to SmallRye Mutiny Uni/Multi

### DISCUSSION THREAD

**Discussion by Stwissel** [quarkusio/quarkus#44670] [Q&A] [answered]
I’m very comfortable with vert.x’s Future/Promise and fairly new to Mutiny. Looking for a guide along the lines of “When you do this with a future, do that with Uni/Multi”

I frequently have code like:

```java
somethingFutureA()
   .compose(this::takeAmakeB)
   .compose(this::takeBmakeC)
   .onSuccess(ctx::end);
```

Or

```java
List<Future<?>> allTheThings = kickoffAllThings();
Future.all(allTheThings)
  .onSuccess(result -> {…})
  .onFailure(ShitHappens::noFuture);
```

Methods I design often look like this

```java
Future<SomeResult> attemptSomeResult(SomeData parameter) {
   Promise<SomeResult> outcome = Promise.promise();
   // do stuff here
   if (success) {
      promise.complete(aSomeResultObject);
   } else {
      promise.fail(errorCondition);
   }
  return Promise.future();
}
```

What would be the Mutiny way?

[a comment by an automated account is not shown]

**Comment by Stwissel** [comment]
So far my impression:

`onSuccess, onFailure, onComplete` -> `subscribe` 
`compose, map` -> `onItem` 
`promise` -> `createFrom` 
`Future.all` -> `Uni.join` 

But I’m sure there are sufficient subtleties to trip over. I’m especially curious about how to implement `takeAmakeB` style Uni (or learn Mutiny style)

Subscribe to a Multi looks very much like an Observable/Observer. I wonder how could multiple subscribers subscribe to a Multi, receiving the same items

**Reply by jponge** [reply, marked as answer]
Coming from Vert.x the main difference of a `Uni` vs a `Future` is that a `Uni` needs to be subscribed to before the actual action is triggered, whereas you can't control when a Vert.x future starts. You will find lots of composition operations to transform values, compose asynchronous results, handle errors, etc.

You might want to start with https://smallrye.io/smallrye-mutiny/latest/tutorials/hello-mutiny/ and look over the documentation here.

You might also look at the samples in https://github.com/smallrye/smallrye-mutiny/tree/main/workshop-examples

Hope it helps!
